core/services/reaper.go

```diff
@@ -3,30 +3,107 @@ package services
 import (
 	"time"
 
+	"github.com/pkg/errors"
 	"github.com/smartcontractkit/chainlink/core/logger"
 	"github.com/smartcontractkit/chainlink/core/store"
 	"github.com/smartcontractkit/chainlink/core/store/orm"
 	"github.com/smartcontractkit/chainlink/core/utils"
 )
 
-type storeReaper struct {
-	store  *store.Store
-	config orm.ConfigReader
+type StoreReaper struct {
+	store           *store.Store
+	config          *orm.Config
+	highestSeenHead HighestSeenHeadFunc
+	log             *logger.Logger
 }
 
-// NewStoreReaper creates a reaper that cleans stale objects from the store.
-func NewStoreReaper(store *store.Store) utils.SleeperTask {
-	return utils.NewSleeperTask(&storeReaper{
-		store:  store,
-		config: store.Config,
-	})
+// NewStoreReaper instantiates a new store reaper object
+func NewStoreReaper(store *store.Store, highestSeenHead HighestSeenHeadFunc) *StoreReaper {
+	return &StoreReaper{
+		store:           store,
+		config:          store.Config,
+		highestSeenHead: highestSeenHead,
+		log:             logger.CreateLogger(logger.Default.With("id", "reaper")),
+	}
+}
+
+// NewStoreReaperTask creates a reaper sleeper task that cleans stale objects from the store.
+func NewStoreReaperTask(store *store.Store, highestSeenHead HighestSeenHeadFunc) utils.SleeperTask {
+	return utils.NewSleeperTask(NewStoreReaper(store, highestSeenHead))
 }
 
-func (sr *storeReaper) Work() {
-	recordCreationStaleThreshold := sr.config.ReaperExpiration().Before(
-		sr.config.SessionTimeout().Before(time.Now()))
+// Work executes when the sleepertask is woken up
+func (sr *StoreReaper) Work() {
+	recordCreationStaleThreshold := sr.config.ReaperExpiration().Before(sr.config.SessionTimeout().Before(time.Now()))
 	err := sr.store.DeleteStaleSessions(recordCreationStaleThreshold)
 	if err != nil {
-		logger.Error("unable to reap stale sessions: ", err)
+		sr.log.Error("Reaper: unable to reap stale sessions: ", err)
+	}
+	err = sr.ReapEthTxes()
+	if err != nil {
+		sr.log.Error("Reaper: unable to reap old eth_txes: ", err)
 	}
 }
+
+// ReapEthTxes deletes old eth_txes
+func (sr *StoreReaper) ReapEthTxes() error {
+	threshold := sr.config.EthTxReaperThreshold()
+	if threshold == 0 {
+		sr.log.Debug("Reaper: ETH_TX_REAPER_THRESHOLD set to 0; skipping reapEthTxes")
+		return nil
+	}
+	latestHead := sr.highestSeenHead()
+	if latestHead == nil {
+		sr.log.Debug("Reaper: reapEthTxes got nil head; skipping")
+		return nil
+	}
+	headNum := latestHead.Number
+	minBlockNumberToKeep := headNum - int64(sr.config.EthFinalityDepth())
+	mark := time.Now()
+	ageThreshold := mark.Add(-threshold)
+	db := sr.store.DB
+
+	sr.log.Infof("Reaper: reaping old eth_txes created before %s", ageThreshold.Format(time.RFC3339))
+
+	// Delete old confirmed eth_txes
+	err := orm.Batch(func(_, limit uint) (count uint, err error) {
+		res := db.Exec(`
+WITH old_enough_receipts AS (
+	SELECT tx_hash FROM eth_receipts
+	WHERE block_number < ?
+	ORDER BY block_number ASC, id ASC
+	LIMIT ?
```

Comment:
hm wouldn't this be returning the same eth_receipts if there are more than 1000 (unless we are also deleting them somewhere?)

Reply:
Good point, but the foreign key covers us. I made a comment explaining this, since it's non-obvious

Reply:
ah great 👍🏼